osu.Game.Rulesets.Taiko/Mods/TaikoModClassic.cs

```diff
@@ -1,22 +1,71 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Diagnostics;
+using osu.Framework.Graphics;
+using osu.Framework.Bindables;
+using osu.Framework.Extensions.EnumExtensions;
+using osu.Game.Beatmaps.Legacy;
 using osu.Game.Rulesets.Mods;
 using osu.Game.Rulesets.Taiko.Objects;
 using osu.Game.Rulesets.Taiko.UI;
 using osu.Game.Rulesets.UI;
+using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Taiko.Objects.Drawables;
 
 namespace osu.Game.Rulesets.Taiko.Mods
 {
-    public class TaikoModClassic : ModClassic, IApplicableToDrawableRuleset<TaikoHitObject>, IUpdatableByPlayfield
+    public class TaikoModClassic : ModClassic, IApplicableToDrawableRuleset<TaikoHitObject>, IApplicableToDrawableHitObject, IUpdatableByPlayfield
     {
+        /// <summary>
+        /// The maximum aspect ratio with classic mod enabled.
+        /// </summary>
+        private const float classic_max_aspect_ratio = 2.0f / 1.0f;
+
+        /// <summary>
+        /// The maximum time range with classic mod enabled. This will be enforced in different ways depending on the
+        /// <see cref="LegacyMods"/> enabled.
+        /// </summary>
+        private static readonly double classic_max_time_range = DrawableTaikoRuleset.AspectRatioToTimeRange(classic_max_aspect_ratio);
+
+        /// <summary>
+        /// The classic hidden aspect ratio. Note that time rate is also stretched to this from the default aspect ratio.
+        /// </summary>
+        private const float classic_hidden_aspect_ratio = 4.0f / 3.0f;
+
+        private LegacyMods enabledMods = LegacyMods.None;
+
         private DrawableTaikoRuleset? drawableTaikoRuleset;
 
+        private readonly Bindable<float> hiddenFadeOutDuration = new Bindable<float>(0.4f);
+
+        private readonly Bindable<float> hiddenInitialAlpha = new Bindable<float>(0.65f);
+
+        /// <summary>
+        /// Mark a legacy mod as enabled. This class will then apply those mod's appropriate classic adjustments to the
+        /// ruleset. These adjustments are aimed to match stable's behaviour.
+        /// </summary>
+        public void EnableLegacyMods(LegacyMods legacyMods)
+        {
+            enabledMods = enabledMods | legacyMods;
+        }
```

Comment:
Also, on technical grounds, the way that the mod interdependency is implemented via this method feels very wrong. Not only is this using `LegacyMods` for whatever reason, it also requires each mod that needs to have its logic changed depending on another mod to self-report. Both feel like very artificial choices.

Reply:
A better alternative might be to make `DrawableTaikoRuleset` pass in all enabled mods. `LegacyMods` is used for ease of mod (combination) detection, the alternative to that would be a list of boolean, or doing a list search to detect the existence of a mod each time.

Reply:
Passing in the list of mods and using a standard containment checks is preferable to me as it helps contain the spread of legacy logic in other classes.

Reply:
Makes sense to avoid other mods having to deal with classic explicitly. List search should work fine too.

Reply:
Applied in 9b7158822b7255320cc55f8ec2026b527197e03f. A nice bonus with this is `DrawableTaikoRuleset` doesn't have to call `TaikoModClassic` now, as `TaikoModClassic` can fetch the mods from `DrawableTaikoRuleset`.